File: java/leetcode_grind/Day632.java
package leetcode_grind;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class Day632 {
    // https://leetcode.com/problems/regions-cut-by-slashes/description/?envType=daily-question&envId=2024-08-10
    static class Solution1 {
        private static final int[][] DIRECTIONS = {
                { 0, 1 },
                { 0, -1 },
                { 1, 0 },
                { -1, 0 },
        };

        public int regionsBySlashes(String[] grid) {
            int gridSize = grid.length;
            int[][] expandedGrid = new int[gridSize * 3][gridSize * 3];

            for (int i = 0; i < gridSize; i++) {
                for (int j = 0; j < gridSize; j++) {
                    int baseRow = i * 3;
                    int baseCol = j * 3;

                    if (grid[i].charAt(j) == '\\') {
                        expandedGrid[baseRow][baseCol] = 1;
                        expandedGrid[baseRow + 1][baseCol + 1] = 1;
                        expandedGrid[baseRow + 2][baseCol + 2] = 1;
                    } else if (grid[i].charAt(j) == '/') {
                        expandedGrid[baseRow][baseCol + 2] = 1;
                        expandedGrid[baseRow + 1][baseCol + 1] = 1;
                        expandedGrid[baseRow + 2][baseCol] = 1;
                    }
                }
            }

            int regionCount = 0;
            for (int i = 0; i < gridSize * 3; i++) {
                for (int j = 0; j < gridSize * 3; j++) {
                    if (expandedGrid[i][j] == 0) {
                        floodFill(expandedGrid, i, j);
                        regionCount++;
                    }
                }
            }
            return regionCount;
        }

        void floodFill(int[][] expandedGrid, int row, int col) {
            Queue<int[]> queue = new LinkedList<>();
            expandedGrid[row][col] = 1;
            queue.add(new int[] { row, col });

            while (!queue.isEmpty()) {
                int[] currentCell = queue.poll();
                for (int[] direction : DIRECTIONS) {
                    int newRow = direction[0] + currentCell[0];
                    int newCol = direction[1] + currentCell[1];

                    if (isValidCell(expandedGrid, newRow, newCol)) {
                        expandedGrid[newRow][newCol] = 1;
                        queue.add(new int[] { newRow, newCol });
                    }
                }
            }
        }

        boolean isValidCell(int[][] expandedGrid, int row, int col) {
            int n = expandedGrid.length;
            return (row >= 0 &&
                    col >= 0 &&
                    row < n &&
                    col < n &&
                    expandedGrid[row][col] == 0);
        }
    }

    static class Solution2 {
        public int regionsBySlashes(String[] grid) {
            int gridSize = grid.length;
            int totalTriangles = gridSize * gridSize * 4;
            int[] parentArray = new int[totalTriangles];
            Arrays.fill(parentArray, -1);

            int regionCount = totalTriangles;

            for (int row = 0; row < gridSize; row++) {
                for (int col = 0; col < gridSize; col++) {
                    if (row > 0) {
                        regionCount -= unionTriangles(
                                parentArray,
                                getTriangleIndex(gridSize, row - 1, col, 2),
                                getTriangleIndex(gridSize, row, col, 0));
                    }
                    if (col > 0) {
                        regionCount -= unionTriangles(
                                parentArray,
                                getTriangleIndex(gridSize, row, col - 1, 1),
                                getTriangleIndex(gridSize, row, col, 3));
                    }

                    if (grid[row].charAt(col) != '/') {
                        regionCount -= unionTriangles(
                                parentArray,
                                getTriangleIndex(gridSize, row, col, 0),
                                getTriangleIndex(gridSize, row, col, 1));
                        regionCount -= unionTriangles(
                                parentArray,
                                getTriangleIndex(gridSize, row, col, 2),
                                getTriangleIndex(gridSize, row, col, 3));
                    }

                    if (grid[row].charAt(col) != '\\') {
                        regionCount -= unionTriangles(
                                parentArray,
                                getTriangleIndex(gridSize, row, col, 0),
                                getTriangleIndex(gridSize, row, col, 3));
                        regionCount -= unionTriangles(
                                parentArray,
                                getTriangleIndex(gridSize, row, col, 2),
                                getTriangleIndex(gridSize, row, col, 1));
                    }
                }
            }
            return regionCount;
        }

        int getTriangleIndex(int gridSize, int row, int col, int triangleNum) {
            return (gridSize * row + col) * 4 + triangleNum;
        }

        int unionTriangles(int[] parentArray, int x, int y) {
            int parentX = findParent(parentArray, x);
            int parentY = findParent(parentArray, y);

            if (parentX != parentY) {
                parentArray[parentX] = parentY;
                return 1;
            }
            return 0;
        }

        int findParent(int[] parentArray, int x) {
            if (parentArray[x] == -1)
                return x;
            return parentArray[x] = findParent(parentArray, parentArray[x]);
        }
    }
}

File: src/leetcode_grind/day_632.rs
// https://leetcode.com/problems/regions-cut-by-slashes/description/?envType=daily-question&envId=2024-08-10
pub fn regions_by_slashes(grid: Vec<String>) -> i32 {
    fn find(parent_array: &mut Vec<i32>, node: i32) -> i32 {
        if parent_array[node as usize] == -1 {
            return node;
        }
        parent_array[node as usize] = find(parent_array, parent_array[node as usize]);
        parent_array[node as usize]
    }

    fn union(parent_array: &mut Vec<i32>, node1: usize, node2: usize) -> i32 {
        let parent1 = find(parent_array, node1 as i32);
        let parent2 = find(parent_array, node2 as i32);
        if parent1 == parent2 {
            return 1;
        }

        parent_array[parent2 as usize] = parent1;
        0
    }

    let grid_size = grid.len();
    let grid = grid
        .into_iter()
        .map(|x| x.chars().clone().collect::<Vec<_>>())
        .collect::<Vec<Vec<_>>>();
    let points_per_side = grid_size + 1;
    let total_points = points_per_side * points_per_side;

    let mut parent_array = vec![-1; total_points];

    for i in 0..points_per_side {
        for j in 0..points_per_side {
            if i == 0 || j == 0 || i == points_per_side - 1 || j == points_per_side - 1 {
                let point = i * points_per_side + j;
                parent_array[point] = 0;
            }
        }
    }

    parent_array[0] = -1;
    let mut region_count = 1;

    for i in 0..grid_size {
        for j in 0..grid_size {
            if grid[i][j] == '/' {
                let top_right = i * points_per_side + (j + 1);
                let bottom_left = (i + 1) * points_per_side + j;
                region_count += union(&mut parent_array, top_right, bottom_left);
            } else if grid[i][j] == '\\' {
                let top_left = i * points_per_side + j;
                let bottom_right = (i + 1) * points_per_side + (j + 1);
                region_count += union(&mut parent_array, top_left, bottom_right);
            }
        }
    }

    region_count
}
